### PR TITLE
Export default Blender scene as a RAMSES scene

### PR DESCRIPTION
**Pull request by dwlsalmeida**
This is a bit rough and bloated still, but nevertheless apparently functional. I have not yet tried to visualize it with the RAMSES viewer to see if it actually renders correctly but I'll add this commit tomorrow. The validation report is empty, and that is good.

I have taken a few steps to produce some debug messages so we are all on the same page on what operations are being done. I have attached what is printed for the default Blender scene - which is comprised of a cube, a point light and a perspective camera here: [debug.txt](https://github.com/GENIVI/ramses-blender/files/3270154/debug.txt).

The code base could see some refactoring to comply more with PEP8 or something, this has not gone unnoticed. Right now I am just trying to get a minimally functional scene to display so we can more carefully build from there.

Unfortunately I have not figured out yet why there seems to be some caching going on on Blender's side. This breaks everything on the second run since all the data structures get duplicated. Maybe somebody has a better take on this issue?

@gunnarx What do you think?

I am also specially looking for feedback on whether render passes and render groups were done right, mainly taking into account that there might be multiple cameras and multiple view layers per Blender scene. 

Lastly, maybe we should add a few tests? I want to be sure things do not break going forward.

### CODE REVIEW

File: intermediary_representation.py [hunk 1]
```diff
@@ -216,13 +216,33 @@ def __init__(self, scene: bpy.types.Scene, root: Node = Node()):
...
+    def add_node(self, o: bpy.types.Object = None, parent: Node = None) -> Node:
         log.debug(f"Adding Blender Object: {o} as a node for root {self.root} and {parent} as parent")
 
-        node = None
+        node = self._translate(o) if o else Node('Placeholder node')
+
+        if self.root is None:
+            self.root = node
+        else:
+            node_parent = parent if parent else self._resolve_parenting(o)
```
Comment (gunnarx) [inline intermediary_representation.py]: Can the call to _resolve_parenting fail?

File: intermediary_representation.py [hunk 2]
```diff
@@ -216,13 +216,33 @@ def __init__(self, scene: bpy.types.Scene, root: Node = Node()):
...
+        Arguments:
+            o {bpy.types.Object} -- The object to be translated
+
+        Returns:
+            Node -- The node / node hierarchy
+        """
 
         # See https://docs.blender.org/manual/en/dev/editors/3dview/object/types.html
         # See also https://docs.blender.org/manual/en/dev/editors/3dview/object/index.html
+        node = None
```
Comment (gunnarx) [inline intermediary_representation.py]: Assigning a default value takes care of the case if o.type is neither MESH, nor CAMERA.

Personally I prefer assigning the value in an else clause, because it forces us to to consider what it means that o.type is neither MESH nor CAMERA.  Is it an unexpected error - should it be reported to developer, or is it a recoverable error (do something) and/or should it perhaps be logged.

A missing "else" after if/elseif is usually a warning sign, similar to a missing "default" in a switch.   Deal with this case somehow - by some error recovery step, or just reporting it, or if truly nothing should be done, explain that in a comment.

Comment (gunnarx) [reply intermediary_representation.py]: Oh sorry, I had not expanded all the code below it...  Anyway, you can consider the principle.

**Comment by gunnarx** [comment]
I'm not very familiar with either Blender or RAMSES frameworks so it is difficult to comment on the functionality.  I could only add some thoughts that came up in reading the code.

Yes, I would encourage that you add more tests and asserts (I saw some asserts -- good!) since this is a learning process and I think it is easy to miss some details and corner-cases.  So that it will break when bringing in more advanced models that do not behave as expected.  (Or doing this could even help to uncover bugs in **other** programs that you connect to this, even if the bugs are not in this program).

File: intermediary_representation.py (hunk 1 above)
Comment (dwlsalmeida) [reply intermediary_representation.py]: hi there Gunnar!

Actually, now that you have said so, it can create a loop if mishandled.

For flexibility reasons, not every node has to be mapped 1:1 to a Blender object and these nodes without a counterpart in Blender were called 'placeholder' nodes. That is the point of having this intermediary representation in the first place : it exists so we can freely adapt Blender's internal scene representation in ways that make more sense to the exporter.

Now if we consider the case in which somebody creates a placeholder node and add it as root - which we actually do, for instance a placeholder node is created to serve as a handle for the IR scene graph - then this will likely work out this way:

1. find_from_blender_object will hopefully return None because it cannot find a node if we don't provide a Blender object.
2. thus self.root will be returned, in which case a self reference will be created.

And i am not 100% sure about (1) to be honest. 

I should an assert for (1). What do you think about (2)?

File: intermediary_representation.py (hunk 2 above)
Comment (dwlsalmeida) [reply intermediary_representation.py]: I think I have covered all light types, so no.

But (1): I might be wrong, and (2): They might add more light types in the future, in which case yes this will silently fail.

Maybe we can standardize a "UnresolvedNode" type? That would push the error elsewhere. The question then becomes what to do with this "UnresolvedNode" when finally translating to RAMSES.

Comment (dwlsalmeida) [reply intermediary_representation.py]: By the way, previously I was raising an exception for nodes I could not translate yet but this was bad for development since I'd have to comment it out until the exporter had support for most Blender nodes. And still, I think breaking for something we cannot translate is sort of unnecessarily harsh on the end user, maybe @violinyanev can comment on a better 'default' behavior?

**Comment by dwlsalmeida** [comment]
Also something is off I guess.

For the default scene, it works as it should. But if we take a heavy Blender official demo scene, such as [this one](https://download.blender.org/demo/test/classroom.zip) we can check that 891 IR nodes are created but only 304 exist in the database (that is, in the Blender .blend file).

You can check that by breaking on the line 'self.ready_to_translate = True' and comparing these two:
`len([x for x in self.scene_representations[0].graph.traverse()])`
`len(self.scene_representations[0].scene.objects)`

Maybe an assert should be created for this too? It should not be way too off like this.

File: intermediary_representation.py (hunk 2 above)
Comment (gunnarx) [reply intermediary_representation.py]: >But (1): I might be wrong, and (2): They might add more light types in the future, in which case yes this will silently fail.

Yes, exactly, so you see this is about defensive programming even if it seems that the logic is correct or everything is covered.  It is more a principle.  There is something suspicious about a missing "else".  If we are writing a program that is using asserts to alert about unexpected behaviors (which you have done in other places), then there should be something similar in that else clause, so that it does not continue unnoticed.

(In certain cases it could be correct behavior to continue without any action in "else" (or in "default" of a switch/case) but in that case that could also be noted with a comment)

Comment (gunnarx) [reply intermediary_representation.py]: > And still, I think breaking for something we cannot translate is sort of unnecessarily harsh on the end user

For sure, developers and users have different needs.  Sometimes as a developer you are also not happy with the program ending with a fatal error, but sometimes you are.  I think we may have a larger discussion about the various types of error handling strategies.  Maybe best done in a meeting instead of here?

**Comment by gunnarx** [comment]
>Maybe an assert should be created for this too? It should not be way too off like this.

Generally speaking, if it is in conflict with your opinion about how the program should work, then assert it, at least for the moment.  Use asserts to codify your own assumptions about what is correct behavior.  

Again I'm not familiar enough with what you are doing to agree or disagree.  But if you are saying the number of created objects should never be more than the number of objects in the Blender model.  Then assert this statemment.  And then go on to find the error.  You can then keep the assert to trigger if there is a regression at some point.

**Comment by violinyanev** [comment]
I Agree with @gunnarx  here - just put asserts everywhere, whenever you assume something. Add #TODO statements additionally whenever you are unsure or assume the assumption is not correct. Then after we get an initial working setup (for simple scenes - cube, monkey, etc) then we can re-visit the asserts, try more complex scenes, etc.

I wanted to check for possibilities to create unit tests for a blender add-on, but I haven't had the time... I think a good way to do that would be to synchetically create blender data in a unit test, as if blender would create it (or maybe even use existing blender functions, such as "bpy.ops.mesh.primitive_monkey_add", and then run the exporter on them from a script and expect specific ramses objects in the exported scene.

**Comment by violinyanev** [comment]
I have trouble re-starting the addon... I replaced the python files, but blender now fails to load the plugin - says "No module named 'ramses'"... Did you have this issue too?

**Comment by dwlsalmeida** [comment]
> I have trouble re-starting the addon... I replaced the python files, but blender now fails to load the plugin - says "No module named 'ramses'"... Did you have this issue too?

No, but I did not run CMake yet, so the error is probably there? Did you copy the .dll file?

**Comment by dwlsalmeida** [comment]
@violinyanev I am basing 89bce76 on glTF-Blender-IO and not even this can get rid of the caching. This caching duplicates all the state, breaking the exporter and seems to be Python related instead of a fault of Blender itself.

I will investigate further but know I am also stuck on this.

I have also added a few asserts and we correctly crash on these due to the above and possibly other mistakes too.

**Comment by violinyanev** [comment]
Ok, seems to work now! F8 seems to work with the imports reloading

**Comment by dwlsalmeida** [comment]
Guys, I think I have managed to sort out that 'bug' we had talked about earlier. It's seems that's just how it works, since the Python interpreter isn't restarted at every run of the add-on. It seems to be tied to the current Blender process, so it only dies when Blender dies.

Does anybody have something against using the excerpt below in `execute`? Seems to work!
` import importlib`
`import sys`
`importlib.reload(sys.modules[__name__]) # Pristine state @ the next run.`
`return {'FINISHED'}`

**Comment by gunnarx** [comment]
I have no problem with the solution.  But it it sounds like this should be an issue with all programs that use this python API to Blender?  Do you agree?  

If we think so, it is usually a good idea to talk with the developers about it.  It could help them to know about a problem, or they could have another solution that we missed. Maybe they have (or should have) some kind of init/reset function to call?
